src/scripts/autocomplete.js

```diff
@@ -0,0 +1,509 @@
+  var selected = this.el.querySelector("li[data-pid='" + pid + "']"),
+    name = (!!selected) ? selected.textContent : "";
+
+  return {
+    el: selected,
+    pid: pid,
+    name: name
+  };
+};
+
+ProjectAutoComplete.prototype.setProjectBullet = function (pid, tid, el) {
+  var project,
+    className,
+    id = parseInt(pid, 10),
+    elem = el || this.placeholderItem.querySelector(".project-bullet"),
+    result,
+    task;
+
+  if (!!pid) {
+    project = this.el.querySelector("li[data-pid='" + pid + "']");
+    if (id !== 0) {
+      if (!!project) {
+        className = project.querySelector(".project-bullet").className;
+        elem.className = className;
+        result = " - " + project.getAttribute("title");
+        task = project.querySelector("li[data-tid='" + tid + "']");
+        if (!!task) {
+          result += " . " + task.getAttribute("title");
+        }
+        return result;
+      }
+    }
+  }
+  elem.className = "project-bullet";
+  return "";
+};
+
+ProjectAutoComplete.prototype.generateLabel = function (select, id, type, tid) {
+  var selected = false,
+    client,
+    result = "",
+    task;
+
+  select = this.getSelectedProjectByPid(id);
+
+  if (!select) {
+    select = this.getSelected();
+  }
+
+  if (!!select && !!select.el) {
+    selected = select.pid;
+    task = select.el.querySelector(".selected-task") || select.el.querySelector("li[data-tid='" + tid + "']");
+    if (!!task) {
+      result += task.getAttribute("title") + " . ";
+    }
+    client = select.el.parentNode.querySelector(".client-row");
+    if (!!client) {
+      result = client.textContent + " - ";
+    }
+    result += select.el.getAttribute("title");
+  }
+
+  if (parseInt(id, 10) === 0 || !selected) {
+    return "Add " + type;
+  }
+  return result;
+};
+
+ProjectAutoComplete.prototype.filterSelection = function () {
+  var key,
+    that = this,
+    val = that.filter.value.toLowerCase(),
+    row,
+    text;
+
+  if (val === that.lastFilter) {
+    return;
+  }
+
+  if (val.length > 0 && !that.el.classList.contains("filtered")) {
+    that.el.classList.add("filtered");
+  }
+  if (val.length === 0) {
+    that.clearFilters();
+    return;
+  }
+  that.lastFilter = val;
+  that.exactMatch = false;
+  for (key in that.listItems) {
+    if (that.listItems.hasOwnProperty(key)) {
+      row = that.listItems[key];
+      text = row.getAttribute("title").toLowerCase();
+      if (text.indexOf(val) !== -1) {
+        if (text === val) {
+          that.exactMatch = val;
+        }
+        row.classList.add("filter");
+
+        if (row.classList.contains("project-row")) {
+          row.parentNode.classList.add("filter");
+          row.parentNode.parentNode.classList.add("filter");
+        }
+        if (row.classList.contains("client-row")) {
+          row.parentNode.classList.add("filter-match");
+        }
+
+        if (row.classList.contains("task-item")) {
+          row.parentNode.parentNode.classList.add("filter");
```

Comment:
In the future, it would be ideal in cases like `.parentNode` to comment with what e.g. row.parentNode.parentNode actually is. (e.g. `// .project-row`)

The combination of multiple `.parentNode` calls and the non-symmetrical classList adding and removing in this function means in practice a reader doesn't have a great chance of understanding exactly what's happening here. Without help, anyway.